telemetry: add doc for telemetry upgrade tests (#61543)
ceph/ceph: src/pybind/mgr/telemetry/tests/telemetry_upgrade_tests.md

The `upgrade` suite is used to verify that upgrades from the (X-2) releases of
Ceph (where X stands for the current release, i.e tentacle as of writing this

> documentation) to the current stable Ceph release  (_squid_) can complete

Author: thank you! done

> successfully without disruption any ongoing workloads.

Author: thank you! done

There are two goals of this suite:

> - Test that the ceph cluster has been successfully upgraded to the desired

Author: thank you! done

  version without the workloads (eg: basic rados api tests, import/export cli
  tests for rbd, stress testing etc) getting disrupted

> - Test Telemetry works after upgrade

Author: thank you! done

> The below is the upgrade test directory from the squid release branch. Each

Author: thank you! done

release branch upgrade directory includes X-2 upgrade testing. That means, we
can test the upgrade from 2 previous releases to the current one.

```
upgrade
├── reef-x
│   ├── filestore-remove-check
│   ├── parallel
│   │   ├── 0-start.yaml
│   │   ├── 1-tasks.yaml
│   │   ├── upgrade-sequence.yaml
│   │   └── workload
│   └── stress-split
|
├── squid-x
│   ├── parallel
│   │   └── workload
│   └── stress-split
|
└── telemetry-upgrade
    ├── quincy-x

>     └── reef-x

Author: thank you! done

```

> It is possible to test the upgrades from Reef (X-2) or from Squid (X-1) to Tenatacle (X).

Author: thank you! done

> - The `upgrade/reef-x/parallel` and `upgrade/squid-x/parallel` sub-suite
>   installs a quincy or reef cluster, then upgrade the cluster to squid. In
>   parallel, some workloads are run against the cluster, including telemetry
>   workunits.
> - The `upgrade/telemetry-upgrade` sub-suite is identical to
>   `upgrade/reef-x/parallel` and `upgrade/squid-x-parallel` sub-suites above,
>   but these only tests the telemetry workunits and do not run any other
>   workloads. 

Reviewer: ```suggestion
- The `upgrade/quincy-x/parallel` and `upgrade/reef-x/parallel` sub-suite
  installs a Quincy or Reef cluster, then upgrades the cluster to Squid (X). In
  parallel, some workloads are run against the cluster, including telemetry
  workunits.
- The `upgrade/telemetry-upgrade` sub-suite is identical to
  `upgrade/quincy-x/parallel` and `upgrade/reef-x-parallel` sub-suites above,
  but these only test the telemetry workunits and do not run any other
  workloads. 
```

Author: thank you! done

> A simple upgrade test contains the test in the order:

Reviewer: ```suggestion
A simple upgrade test contains these steps in order, divided into separate yaml files:
```

Author: thank you! done

```
├── 0-start.yaml
├── 1-tasks.yaml
├── upgrade-sequence.yaml
└── workload
```

- `0-start.yaml`: This file contains the information about the ceph cluster
  configuration (number of osds, monitors etc) for the test
- `1-tasks.yaml`: This file contains the information of the tasks we want to run
  on the cluster. It is here that we install an older release, then begin
  running the given `workload` and `upgrade-sequence` in parallel.
- `upgrade-sequence.yaml`: This file contains the steps for upgrading the
  cluster to the latest release
- `workloads`: A set of yaml file with workloads we want to run while upgrade is
  in progress

```
- print: "**** done start parallel"
- parallel:
    - workload
    - upgrade-sequence
- print: "**** done end parallel"
```

The `workload` directory contains the workload yaml files just like any other

> suite and the `upgrade-sequence` is responsible for running the upgrade and

Reviewer: ```suggestion
suite and the `upgrade-sequence` is responsible for initiating the upgrade and
```

Author: thank you! done

waiting for it to complete.

```
# renamed tasks: to upgrade-sequence:
upgrade-sequence:
   sequential:
   - print: "**** done start upgrade, wait"
   ...
       mon.a:
         - ceph config set global log_to_journald false --force
         - ceph orch upgrade start --image quay.ceph.io/ceph-ci/ceph:$sha1
         - while ceph orch upgrade status | jq '.in_progress' | grep true && ! ceph orch upgrade status | jq '.message' | grep Error ; do ceph orch ps ; ceph versions ; ceph orch upgrade status ; sleep 30 ; done
   ...
   - print: "**** done end upgrade, wait..."
```

## Telemetry Upgrade tests

The telemetry upgrade test suite verifies that the telemetry is emitting the

> correct collections after the upgrade. This integration test coverage are done

Reviewer: ```suggestion
correct collections after the upgrade. This integration test coverage is done
```

Author: thank you! done

via workunits. Workunits are basically bash scripts that run commands against a
Ceph Cluster. 

> In the same manner as the upgrade test, each release branch `qa/workunits`
> directory includes telemetry test for X-2 releases. That means, we can test
> telemetry before and after the upgrade from previous two releases to the current
> one.

Reviewer: ```suggestion
In the same manner as the `upgrade/parallel` tests, each release branch references the `qa/workunits`
directory, which includes telemetry bash scripts for X-2 releases. That means we can test
telemetry before and after the upgrade from previous two releases to the current
one.
```

Author: thank you! done

For instance, the relevant telemetry workunits for `tentacle` release are:
```
qa/workunits
├── test_telemetry_reef.sh
├── test_telemetry_reef_x.sh
├── test_telemetry_squid.sh
└── test_telemetry_squid_x.sh
```

- `test_telemetry_reef.sh`, tests the presence of telemetry collection on a reef cluster before the upgrade.
- `test_telemetry_reef_x.sh`, tests the presence of new telemetry collection on the X-version cluster after it has been upgrade from reef.
- `test_telemetry_squid.sh`, tests the presence of telemetry collection on a squid cluster before the upgrade

> - `test_telemetry_squid_x.sh`, tests the presence of new telemetry collection on the X-version cluster after it has been upgrade from squid.

Reviewer: Update the workunits referenced here to align with the squid example (see https://github.com/ceph/ceph/tree/squid/qa/workunits)

Author: thank you! done

A sample telemetry upgrade test file contains the following test: 
```
...
# Assert that new collections are available
COLLECTIONS=$(ceph telemetry collection ls)
NEW_COLLECTIONS=("perf_perf" "basic_mds_metadata" "basic_pool_usage"
                 "basic_rook_v01" "perf_memory_metrics" "basic_pool_options_bluestore")
for col in ${NEW_COLLECTIONS[@]}; do
    if ! [[ $COLLECTIONS == *$col* ]];
    then
        echo "COLLECTIONS does not contain" "'"$col"'."
	exit 1
    fi
done
...
```

These workunits are used in the `upgrade` suite, specifically in:
- [upgrade/reef-x/parallel](https://github.com/ceph/ceph/tree/main/qa/suites/upgrade/reef-x/parallel)
- [upgrade/squid-x/parallel](https://github.com/ceph/ceph/tree/main/qa/suites/upgrade/squid-x/parallel)

> - [upgrade/telemetry-upgrade](https://github.com/ceph/ceph/tree/main/qa/suites/upgrade/telemetry-upgrade)

Reviewer: Same as above; update to reflect the squid example (see https://github.com/ceph/ceph/tree/squid/qa/workunits)

Author: thank you! done

```

upgrade
├── reef-x
│   ├── parallel
│   │   └──  1-tasks.yaml
├── squid-x
│   ├── parallel
│   │   └── 1-tasks.yaml
└── telemetry-upgrade
    ├── quincy-x
    └── reef-x

> ```

Reviewer: Same as above; update to reflect squid example.

Author: thank you! done

The `upgrade/reef-x/parallel` and `upgrade/squid-x/parallel` sub-suite installs
a reef or squid cluster, then upgrade the cluster to tentacle. In parallel, some
workloads are run against the cluster, including telemetry workunits. The

> `1-tasks.yaml` file is the place where the workunits are run.

Reviewer: Update to be consistent with the squid example (see https://github.com/ceph/ceph/tree/squid/qa/suites/upgrade)

Author: thank you! done

```
...
- print: "**** done start telemetry quincy..."
- workunit:
    clients:
      client.0:
        - test_telemetry_quincy.sh
- print: "**** done end telemetry quincy..."

- print: "**** done start parallel"
- parallel:
    - workload
    - upgrade-sequence
- print: "**** done end parallel"

- print: "**** done start telemetry x..."
- workunit:
    clients:
      client.0:
        - test_telemetry_quincy_x.sh
- print: "**** done end telemetry x..."
```

The `test_telemetry_quincy.sh` workunit is run on the quincy cluster before the
upgrade and `test_telemetry_quincy_x.sh` is run on the X-version cluster (in
this example `squid`) after the upgrade.

The `upgrade/telemetry-upgrade` sub-suite is identical to the
`upgrade/quincy-x/parallel` and `upgrade/reef-x/parallel` suites as above, but
these tests ONLY test the telemetry workunits and do not run with any other
workloads.

So `upgrade/telemetry-upgrade` is nice to schedule when you just want to verify
that the telemetry workunits are working as expected (they complete much
faster). The `upgrade/[quincy|reef]-x/parallel` suites are nice to schedule if
you want to verify that the workunits are working fine with all the other

> workloads also running.

Reviewer: Nicely said!

> The `upgrade` suite is used to verify that upgrades from the (X-2) releases of
> Ceph (where X stands for the current release, i.e tentacle as of writing this
> documentation) to the current stable Ceph release (_tentacle_) can complete
> successfully without disrupting any ongoing workloads.

Reviewer: ```suggestion
The `upgrade` suite is used to verify that upgrades can complete
successfully without disrupting any ongoing workloads.
```

Let's simplify this introduction since you explain the X-2 aspect a bit later in the doc.

> There are two goals of this suite:
> - Test that the Ceph cluster has been successfully upgraded to the desired
>   version without the workloads (eg: basic rados api tests, import/export cli
>   tests for rbd, stress testing etc) getting disrupted
> - Test Telemetry functions as expected after the upgrade

Reviewer: ```suggestion
```
Let's remove this section since at this stage, you're talking about the upgrade suite as a whole, and this section specifically describes the goals of the telemetry-upgrade sub-suite. Since you talk about these aspects later in the doc, it's better to remove it here.

The diagram below represents the upgrade test directory from the squid release
branch. Each release branch upgrade directory includes X-2 upgrade testing. That
means, we can test the upgrade from 2 previous releases to the current one.

```
upgrade
├── quincy-x
│   ├── filestore-remove-check
│   ├── parallel
│   │   ├── 0-start.yaml
│   │   ├── 1-tasks.yaml
│   │   ├── upgrade-sequence.yaml
│   │   └── workload
│   └── stress-split
|
├── reef-x
│   ├── parallel
│   │   └── workload
│   └── stress-split
|
├── squid-p2p
│   ├── squid-p2p-parallel
│   └── squid-p2p-stress-split
|
└── telemetry-upgrade
    ├── quincy-x
    └── reef-x

```

> Based on the above example where X=squid, is possible to test the upgrades from
> Quincy (X-2) or from Reef (X-1) to Squid (X).

Reviewer: ```suggestion
Based on the above example where X=squid, it is possible to test the upgrades from
Quincy (X-2) or from Reef (X-1) to Squid (X).
```

- The `upgrade/quincy-x/parallel` and `upgrade/reef-x/parallel` sub-suite
  installs a Quincy or Reef cluster, then upgrades the cluster to Squid (X). In
  parallel, some workloads are run against the cluster, including telemetry
  workunits.

> - The `upgrade/telemetry-upgrade` sub-suite is identical to
>   `upgrade/quincy-x/parallel` and `upgrade/reef-x-parallel` sub-suites above,
>   but these only test the telemetry workunits and do not run any other
>   workloads.

Reviewer: ```suggestion
- The `upgrade/telemetry-upgrade` sub-suite is identical to
  `upgrade/quincy-x/parallel` and `upgrade/reef-x/parallel` sub-suites above,
  but these only test the telemetry workunits and do not run any other
  workloads.
```

A simple upgrade test contains these steps in order, divided into separate yaml files:
```
├── 0-start.yaml
├── 1-tasks.yaml
├── upgrade-sequence.yaml
└── workload
```

- `0-start.yaml`: This file contains the information about the ceph cluster
  configuration (number of osds, monitors etc) for the test
- `1-tasks.yaml`: This file contains the information of the tasks we want to run
  on the cluster. It is here that we install an older release, then begin
  running the given `workload` and `upgrade-sequence` in parallel.

> - `upgrade-sequence.yaml`: This file contains the steps for upgrading the
>   cluster to the latest release

Reviewer: ```suggestion
- `upgrade-sequence.yaml`: This file contains the steps for upgrading the
  cluster to the designated release
```
In this case, Squid isn't the latest release, but it is the designated release that we are testing the upgrade sequence on.

> - `workloads`: A set of yaml file with workloads we want to run while upgrade is
>   in progress

Reviewer: ```suggestion
- `workloads`: A set of yaml file with workloads we want to run while the upgrade is
  in progress
```

```
- print: "**** done start parallel"
- parallel:
    - workload
    - upgrade-sequence
- print: "**** done end parallel"
```

The `workload` directory contains the workload yaml files just like any other
suite and the `upgrade-sequence` is responsible for initiating the upgrade and
waiting for it to complete.

```
# renamed tasks: to upgrade-sequence:
upgrade-sequence:
   sequential:
   - print: "**** done start upgrade, wait"
   ...
       mon.a:
         - ceph config set global log_to_journald false --force
         - ceph orch upgrade start --image quay.ceph.io/ceph-ci/ceph:$sha1
         - while ceph orch upgrade status | jq '.in_progress' | grep true && ! ceph orch upgrade status | jq '.message' | grep Error ; do ceph orch ps ; ceph versions ; ceph orch upgrade status ; sleep 30 ; done
   ...
   - print: "**** done end upgrade, wait..."
```

## Telemetry Upgrade tests

> The telemetry upgrade test suite verifies that the telemetry is emitting the
> correct collections after the upgrade. This integration test coverage is done
> via workunits. Workunits are basically bash scripts that run commands against a
> Ceph Cluster.

Reviewer: ```suggestion
The telemetry upgrade sub-suite verifies that telemetry is emitting the
correct collections after the upgrade. This integration test coverage is done
via workunits. Workunits are basically bash scripts that run commands against a
Ceph Cluster.
```

In the same manner as the `upgrade/parallel` tests, each release branch
references the `qa/workunits` directory, which includes telemetry bash scripts
for X-2 releases. That means we can test telemetry before and after the upgrade
from previous two releases to the current one.

> For instance, the relevant telemetry workunits for `squid` release are:

Reviewer: ```suggestion
For instance, the relevant telemetry workunits for the `squid` release are:
```

```
qa/workunits
├── test_telemetry_quincy.sh
├── test_telemetry_quincy_x.sh
├── test_telemetry_reef.sh
└── test_telemetry_reef_x.sh
```

- `test_telemetry_quincy.sh`, tests the presence of telemetry collection on a Quincy cluster before the upgrade.

> - `test_telemetry_quincy_x.sh`, tests the presence of new telemetry collection on the X-version cluster after it has been upgrade from Quincy.

Reviewer: ```suggestion
- `test_telemetry_quincy_x.sh`, tests the presence of new telemetry collection on the X-version cluster after it has been upgraded from Quincy.
```

- `test_telemetry_reef.sh`, tests the presence of telemetry collection on a Reef cluster before the upgrade

> - `test_telemetry_reef_x.sh`, tests the presence of new telemetry collection on the X-version cluster after it has been upgrade from Reef.

Reviewer: ```suggestion
- `test_telemetry_reef_x.sh`, tests the presence of new telemetry collection on the X-version cluster after it has been upgraded from Reef.
```

A sample telemetry upgrade test file contains the following test:
```
...
# Assert that new collections are available
COLLECTIONS=$(ceph telemetry collection ls)
NEW_COLLECTIONS=("perf_perf" "basic_mds_metadata" "basic_pool_usage"
                 "basic_rook_v01" "perf_memory_metrics" "basic_pool_options_bluestore")
for col in ${NEW_COLLECTIONS[@]}; do
    if ! [[ $COLLECTIONS == *$col* ]];
    then
        echo "COLLECTIONS does not contain" "'"$col"'."
	exit 1
    fi
done
...
```

These workunits are used in the `upgrade` suite, specifically in:
- [upgrade/quincy-x/parallel](https://github.com/ceph/ceph/blob/squid/qa/suites/upgrade/quincy-x/parallel/1-tasks.yaml)
- [upgrade/reef-x/parallel](https://github.com/ceph/ceph/blob/squid/qa/suites/upgrade/reef-x/parallel/1-tasks.yaml)
- [upgrade/telemetry-upgrade](https://github.com/ceph/ceph/tree/squid/qa/suites/upgrade/telemetry-upgrade)

```

upgrade
├── reef-x
│   ├── parallel
│   │   └──  1-tasks.yaml
├── squid-x
│   ├── parallel
│   │   └── 1-tasks.yaml
└── telemetry-upgrade
    ├── quincy-x
    └── reef-x

```

> The `upgrade/quincy-x/parallel` and `upgrade/reef-x/parallel` sub-suite installs
> a Quincy or Reef cluster, then upgrade the cluster to Squid. In parallel, some
> workloads are run against the cluster, including telemetry workunits. The
> `1-tasks.yaml` file is the place where the workunits are run.

Reviewer: ```suggestion
The `upgrade/quincy-x/parallel` and `upgrade/reef-x/parallel` sub-suite installs
a Quincy or Reef cluster, then upgrades the cluster to Squid. In parallel, some
workloads are run against the cluster, including telemetry workunits. The
`1-tasks.yaml` file is the place where the workunits are run.
```

```
...
- print: "**** done start telemetry quincy..."
- workunit:
    clients:
      client.0:
        - test_telemetry_quincy.sh
- print: "**** done end telemetry quincy..."

- print: "**** done start parallel"
- parallel:
    - workload
    - upgrade-sequence
- print: "**** done end parallel"

- print: "**** done start telemetry x..."
- workunit:
    clients:
      client.0:
        - test_telemetry_quincy_x.sh
- print: "**** done end telemetry x..."
```

The `test_telemetry_quincy.sh` workunit is run on the Quincy cluster before the
upgrade and `test_telemetry_quincy_x.sh` is run on the X-version cluster (in
this example `squid`) after the upgrade.

The `upgrade/telemetry-upgrade` sub-suite is identical to the
`upgrade/quincy-x/parallel` and `upgrade/reef-x/parallel` suites as above, but
these tests ONLY test the telemetry workunits and do not run with any other
workloads.

So `upgrade/telemetry-upgrade` is nice to schedule when you just want to verify
that the telemetry workunits are working as expected (they complete much
faster). The `upgrade/[quincy|reef]-x/parallel` suites are nice to schedule if
you want to verify that the workunits are working fine with all the other

> workloads also running.

Reviewer: Nicely said!
